### PR TITLE
TRIFFHIR-147

### PR DESCRIPTION
**Pull request by makeshcode360**

### CODE REVIEW

File: apps/client/src/app/export/export.component.html [hunk 1]
```diff
@@ -36,6 +36,7 @@ <h1>Export</h1>
             <label>Output Format</label>
             <select class="form-control" [(ngModel)]="options.responseFormat">
               <option value="application/json">JSON</option>
+              <option value="application/msword">MSWORD</option>
```
Comment (seanmcilvenna) [inline apps/client/src/app/export/export.component.html]: The MWord export options are bundled in under the "IG Publisher Package" tab. "MS Word" should be a separate tab entirely, with (in the future) its own export options. Remove the option from the "select" field, and just create a separate tab for "MS Word". When the user selects the tab, onTabChange() will be called and the options.exportFormat should be changed to (a new enum option) "msword".

File: apps/client/src/app/export/export.component.html [hunk 2]
```diff
@@ -77,6 +78,7 @@ <h1>Export</h1>
             <label>Output Format</label>
             <select class="form-control" [(ngModel)]="options.responseFormat" (change)="responseFormatChanged()">
               <option value="application/json">JSON</option>
+              <option value="application/msword">MSWORD</option>
```
Comment (seanmcilvenna) [inline apps/client/src/app/export/export.component.html]: See comment above. MS Word should be separate tab, not an option under "Output Format".

File: apps/client/src/app/export/export.component.ts [hunk 3]
```diff
@@ -198,18 +198,33 @@ export class ExportComponent implements OnInit {
...
+    const extension = this.options.responseFormat === 'application/xml' ? '.xml' : this.options.responseFormat === 'application/msword' ? '.docx' : '.json';
+
+
+    console.log("*****EXTENSION:" + extension);
 
     try {
       switch (this.options.exportFormat) {
         case ExportFormats.HTML:
-          this.exportService.exportHtml(this.options)
+          //Need to discuss the difference between HTML vs Bundle
```
Comment (seanmcilvenna) [inline apps/client/src/app/export/export.component.ts]: There should be a new enum option on ExportFormats for "MSWord". Need a separate "case" block here.

File: apps/server/src/app/export/msword.ts [hunk 4]
```diff
@@ -40,6 +55,59 @@ export class MSWordExporter {
...
+      this.body.push(new Paragraph({
+        text: "Copyright:" + implementationGuide.copyright
+      }));
+
+      this.body.push(new Paragraph({
+        text: ""
+      }));
+
+      bundle.entry
+        .filter(entry => entry.resource && (entry.resource.resourceType === 'StructureDefinition' ||
```
Comment (seanmcilvenna) [inline apps/server/src/app/export/msword.ts]: I suggest creating separate filters for each resourceType. Example:
```
this.body.push(new Paragraph({ text: 'Profiles' });
bundle.entry.filter(entry => entry.resource && entry.resourceType === 'StructureDefinition')
  .forEach(entry => {
    const structureDefinition: StructureDefinition = <StructureDefinition> entry.resource;   // we can do this because we know entry.resource will be a StructureDefinition because of the filter
    this.body.push(new Paragraph({ text: `Name: ${structureDefinition.name` });    // name is required, so we can expect it to be populated with a value
    if (structureDefinition.title) {
      this.body.push(new Paragraph({ text: `Title: ${structureDefinition.title` });
    }

    if (structureDefinition.differential) {
      (structureDefinition.differential.element || []).forEach((element, eIndex) => {
        this.body.push(new Paragraph({ text: `Element ${eIndex + 1}` });
        this.body.push(new Paragraph({ text: `Path: ${element.path}` });
        // more fields...
      });
  });

  // filter and forEach on "ValueSet". output properties specific to value set

  // filter and forEach on "CodeSystem". output properties specific to code system

  // filter and forEach on "CapabilityStatement". output properties specific to capability statement

  // etc.
```

File: apps/client/src/app/export/export.component.html (hunk 1 above)
Comment (makeshcode360) [reply apps/client/src/app/export/export.component.html]: Completed.

File: apps/client/src/app/export/export.component.html (hunk 2 above)
Comment (makeshcode360) [reply apps/client/src/app/export/export.component.html]: Completed

File: apps/client/src/app/export/export.component.ts (hunk 3 above)
Comment (makeshcode360) [reply apps/client/src/app/export/export.component.ts]: Completed

File: apps/server/src/app/export/msword.ts (hunk 4 above)
Comment (makeshcode360) [reply apps/server/src/app/export/msword.ts]: @seanmcilvenna this is where I had issue with undefined properties name and title of StructureDefinition.  Remember, I asked this question couple of days ago. Bottomline is receiving compilation error that name and title are not properties of StructureDefinition.

Comment (seanmcilvenna) [reply apps/server/src/app/export/msword.ts]: with the way the code is now, I can see the errors about name and title... you aren't casting `entry.resource` to anything, so the compiler just thinks it's a DomainResource, which doesn't have any resource-specific properties on it, such as name and title. You have to determine what type of resource it is by checking `entry.resource.resourceType` and then cast `entry.resource` to the appropriate type.

For example:

```
if (entry.resource.resourceType === 'StructureDefinition') {
  const structureDefinition: StructureDefinition = <StructureDefinition> entry.resource;
  ...
}
```

Comment (makeshcode360) [reply apps/server/src/app/export/msword.ts]: Yes, I did cast to StructureDefinition. But, the StructureDefinition does not have these properties defined/inheriting. This is how I casted. 
const structureDefinition = <StructureDefinition> entry.resource;

Comment (seanmcilvenna) [reply apps/server/src/app/export/msword.ts]: there is no strong typing in `const structureDefinition = entry.resource;` 
`structureDefinition` in this case is not type-specific... it's just the variable name. So, it is inheriting whatever the type is that it is set from, which in this case is `entry.resource`.
`entry.resource` is of type `DomainResource`... _not_ of type `StructureDefinition`.

What you need instead is: `const structureDefinition: StructureDefinition = <StructureDefinition> entry.resource;`

Comment (makeshcode360) [reply apps/server/src/app/export/msword.ts]: @seanmcilvenna, I tried this way today morning and received the same error.

Comment (makeshcode360) [reply apps/server/src/app/export/msword.ts]: Let me try again.

Comment (makeshcode360) [reply apps/server/src/app/export/msword.ts]: Same error @seanmcilvenna

Comment (seanmcilvenna) [reply apps/server/src/app/export/msword.ts]: can you send me an updated code snippet?

Comment (seanmcilvenna) [reply apps/server/src/app/export/msword.ts]: or commit your changes in their current state? so I can see them reflected in this PR?

Comment (makeshcode360) [reply apps/server/src/app/export/msword.ts]: @seanmcilvenna I committed the changes with current state. Please check

**Comment by seanmcilvenna** [comment]
Ah... I see the problem.
You are using StructureDefinition from the fhir.js module. That is not the complete representation of a StructureDefinition. It is only the bare minimum needed for the fhir.js module to work.
You should use the StructureDefinition models from `libs/tof-lib/src/lib/r4/fhir.ts` and `libs/tof-lib/src/lib/stu3/fhir.ts`, like so:
`import {StructureDefinition} from '../../../../../libs/tof-lib/src/lib/r4/fhir';`
Additionally, you could use _both_ versions of the models like so:

```
import {StructureDefinition as STU3StructureDefinition} from '../../../../../libs/tof-lib/src/lib/r4/fhir';
import {StructureDefinition as R4StructureDefinition} from '../../../../../libs/tof-lib/src/lib/r4/fhir';

...

const structureDefinition: STU3StructureDefinition | R4StructureDefinition = <STU3StructureDefinition | R4StructureDefinition> entry.resource;
```

For intellisense, this will create a composite of the two models, showing what is common among both versions of the models, and if you wanted version-specific properties, you would have to cast it to the version-specific model.

**Comment by makeshcode360** [comment]
> Ah... I see the problem.
> You are using StructureDefinition from the fhir.js module. That is not the complete representation of a StructureDefinition. It is only the bare minimum needed for the fhir.js module to work.
> You should use the StructureDefinition models from `libs/tof-lib/src/lib/r4/fhir.ts` and `libs/tof-lib/src/lib/stu3/fhir.ts`, like so:
> `import {StructureDefinition} from '../../../../../libs/tof-lib/src/lib/r4/fhir';`
> Additionally, you could use _both_ versions of the models like so:
> 
> ```
> import {StructureDefinition as STU3StructureDefinition} from '../../../../../libs/tof-lib/src/lib/r4/fhir';
> import {StructureDefinition as R4StructureDefinition} from '../../../../../libs/tof-lib/src/lib/r4/fhir';
> 
> ...
> 
> const structureDefinition: STU3StructureDefinition | R4StructureDefinition = <STU3StructureDefinition | R4StructureDefinition> entry.resource;
> ```
> 
> For intellisense, this will create a composite of the two models, showing what is common among both versions of the models, and if you wanted version-specific properties, you would have to cast it to the version-specific model.

**Comment by makeshcode360** [comment]
> > Ah... I see the problem.
> > You are using StructureDefinition from the fhir.js module. That is not the complete representation of a StructureDefinition. It is only the bare minimum needed for the fhir.js module to work.
> > You should use the StructureDefinition models from `libs/tof-lib/src/lib/r4/fhir.ts` and `libs/tof-lib/src/lib/stu3/fhir.ts`, like so:
> > `import {StructureDefinition} from '../../../../../libs/tof-lib/src/lib/r4/fhir';`
> > Additionally, you could use _both_ versions of the models like so:
> > ```
> > import {StructureDefinition as STU3StructureDefinition} from '../../../../../libs/tof-lib/src/lib/r4/fhir';
> > import {StructureDefinition as R4StructureDefinition} from '../../../../../libs/tof-lib/src/lib/r4/fhir';
> > 
> > ...
> > 
> > const structureDefinition: STU3StructureDefinition | R4StructureDefinition = <STU3StructureDefinition | R4StructureDefinition> entry.resource;
> > ```
> > 
> > 
> > For intellisense, this will create a composite of the two models, showing what is common among both versions of the models, and if you wanted version-specific properties, you would have to cast it to the version-specific model.

Sure @seanmcilvenna I will check it out.

**Comment by makeshcode360** [comment]
@seanmcilvenna, other than name all other elements are undefined in structure definition. Also, in the differential elements, I see most of the attributes undefined.

File: apps/server/src/app/export/msword.ts [hunk 5]
```diff
@@ -40,6 +58,225 @@ export class MSWordExporter {
...
+
+          this.body.push(new Paragraph({ text: `Name: ${structureDefinition.name}` }));    // name is required, so we can expect it to be populated with a value
+          if (structureDefinition.title) {
+            this.body.push(new Paragraph({ text: `Title: ${structureDefinition.title}` }));    // name is required, so we can expect it to be populated with a value
+          }
+
+
+          if (structureDefinition.differential) {
+            (structureDefinition.differential.element || []).forEach((element, eIndex) => {
+                if(element.id === 'Extension'){
```
Comment (seanmcilvenna) [inline apps/server/src/app/export/msword.ts]: Why `if (element.id === 'Extension') {`?
Seems to me the logic can be the same regardless of the value of element.id

File: apps/server/src/app/export/msword.ts [hunk 6]
```diff
@@ -40,6 +58,225 @@ export class MSWordExporter {
...
+        );
+
+      bundle.entry
+        .filter(entry => entry.resource && (entry.resource.resourceType === 'CodeSystem'))
+        .forEach((entry) => {
+            const extensionUrlKeys = Object.keys(Globals.extensionUrls);
+            const extensionUrls = extensionUrlKeys.map((key) => Globals.extensionUrls[key]);
+            const keys = Object.keys(entry.resource);
+
+            keys.forEach((key) => {
```
Comment (seanmcilvenna) [inline apps/server/src/app/export/msword.ts]: I'd rather output a select number of properties, rather than just looping through all the keys...
* name
* title
* purpose
* description
* filter
  * filter.code
  * filter.description
* property
  * property.code
  * property.description
* total number of concept

File: apps/server/src/app/export/msword.ts [hunk 7]
```diff
@@ -40,6 +58,225 @@ export class MSWordExporter {
...
+          this.body.push(new Paragraph({
+            text: ""
+          }));
+          }
+        );
+
+      bundle.entry
+        .filter(entry => entry.resource && (entry.resource.resourceType === 'ValueSet'))
+        .forEach((entry) => {
+            const extensionUrlKeys = Object.keys(Globals.extensionUrls);
```
Comment (seanmcilvenna) [inline apps/server/src/app/export/msword.ts]: I'd rather output a select number of properties, rather than just looping through all the keys...
* name
* title
* description
* purpose
* copyright
* compose
  * include
    * system
    * concept
      * code
      * display
    * filter
      * property, op, value
  * exclude
    * system
    * concept
      * code
      * display
    * filter
      * property, op, value

File: apps/server/src/app/export/msword.ts [hunk 8]
```diff
@@ -40,6 +58,225 @@ export class MSWordExporter {
...
+                this.body.push(new Paragraph({ text: `Short: ${element.short}` }));
+                this.body.push(new Paragraph({ text: `Label: ${element.label}` }));
+
+              }
+            );
+          };
+
+
+
+            keys.forEach((key) => {
```
Comment (seanmcilvenna) [inline apps/server/src/app/export/msword.ts]: I like the way the information is output above, showing each important property, rather than just dumping out all the keys. I would suggest removing this. Also, I suggest creating a table for each element like you do here and putting "fixed", "min", "max", "mustSupport", etc. in the table as a individual rows. I also suggest checking the existence of those properties first, so that we don't output empty rows. Tables are a nice way of displaying repeating data, and can be applied to code below this as well for ValueSet and CodeSystem.

**Comment by makeshcode360** [comment]
@seanmcilvenna I checked in the code. I think its definitely closer but may need some tweaks. Please check and let me know your feedback. Thanks
